File: python/pipeql_python/driver.py
"""Zero-boilerplate PipeQL database adapters (``pipeql_python.driver``).

Wraps any DB-API 2.0 (PEP 249) connection -- or an ``asyncpg`` connection --
and auto-handles:

* compilation (PipeQL source -> dialect SQL)
* parameter binding (named PipeQL params -> positional driver args)
* dispatch (SELECT via ``cursor.fetchall()``, mutations via ``cursor.execute``)
* ``$data`` object expansion (partial insert/update with zero boilerplate)
* ``insert_and_fetch`` / ``update_and_fetch`` (single-call write + return)

Supported connections (auto-detected, or forced via ``driver=``):
``sqlite3``, ``duckdb``, ``psycopg`` / ``psycopg2``, ``pymysql`` /
``mysql.connector``, and ``asyncpg``.

Example::

    import sqlite3
    from pipeql_python.driver import create_pipeql_driver

    db = create_pipeql_driver(sqlite3.connect("notes.db"))

    rows = db.query("from notes | filter category == $cat", {"cat": "Ideas"})
    run = db.execute("into notes | insert [title = $title]", {"title": "New Note"})
    note = db.insert_and_fetch("into notes | insert $data", {"title": "New Note"})
    updated = db.update_and_fetch(
        "from notes | filter id == $id | update $data", {"id": 1, "data": {"title": "Renamed"}})
"""

from __future__ import annotations

import re
from typing import Any, Dict, List, Optional, Tuple

from .pipeql_python import compile as _compile

__all__ = ["create_pipeql_driver", "detect_driver", "PipeqlDriver"]

_DIALECT_BY_DRIVER = {
    "sqlite3": "sqlite",
    "duckdb": "duckdb",
    "psycopg": "postgres",
    "asyncpg": "postgres",
    "pymysql": "mysql",
    "mysql": "mysql",
    "dbapi": None,
}

_ASYNC_DRIVERS = frozenset({"asyncpg"})
_NO_RETURNING = frozenset({"pymysql", "mysql"})

_DATA_RE = re.compile(r"\$data\b")
_IDENT_RE = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")
_RETURNING_RE = re.compile(r"\bRETURNING\b", re.IGNORECASE)


def _bind(compiled: Dict[str, Any], params: Optional[Dict[str, Any]]) -> List[Any]:
    """Map named PipeQL params to positional driver args.

    Literal-derived params (concrete type in the analysis param map, or a
    select-mode string literal with no param-map entry) bind as themselves.
    A user ``$param`` (type ``Any``) missing from ``params`` is a bug — fail
    loudly instead of silently binding its name and returning wrong data.
    """
    analysis = compiled.get("analysis") or {}
    types = {p["name"]: p.get("ty") for p in analysis.get("param_map", [])}
    args = []
    for name in compiled["params"]:
        if params is not None and name in params:
            args.append(params[name])
        elif types.get(name) == "Any":
            raise KeyError(
                f"missing value for parameter ${name} — pass it in params, "
                f"e.g. {{ {name}: ... }}"
            )
        else:
            # Literal-derived param: the value is the name itself.
            args.append(name)
    return args


def _rows_as_dicts(cursor) -> List[Dict[str, Any]]:
    cols = [d[0] for d in (cursor.description or [])]
    return [dict(zip(cols, row)) for row in cursor.fetchall()]


def _parse_command_tag(status: str) -> Optional[int]:
    parts = str(status).split()
    if parts and parts[-1].isdigit():
        return int(parts[-1])
    return None


def _expand_data(source: str, data: Any) -> Tuple[str, Dict[str, Any]]:
    """Rewrite ``$data`` in ``source`` into explicit column assignments.

    Values from ``data`` become bound params named ``data0``, ``data1``, ...
    """
    if not isinstance(data, dict):
        raise ValueError(
            f"$data object expansion requires a dict (got {type(data).__name__})"
        )
    if not data:
        raise ValueError("$data object expansion requires at least one property")
    parts = _DATA_RE.split(source)
    chunks = [parts[0]]
    values: Dict[str, Any] = {}
    n = 0
    for i in range(1, len(parts)):
        prev = chunks[-1].rstrip()
        last = prev[-1] if prev else ""
        in_brackets = last in ("[", ",")
        assignments = []
        for key, v in data.items():
            if not _IDENT_RE.match(key):
                raise ValueError(
                    f"cannot expand $data: column {key!r} is not a valid identifier"
                )
            pname = f"data{n}"
            n += 1
            values[pname] = v
            assignments.append(f"{key} = ${pname}")
        body = ", ".join(assignments)
        chunks.append(body if in_brackets else f"[{body}]")
        chunks.append(parts[i])
    return "".join(chunks), values


def _with_returning(sql: str) -> str:
    if _RETURNING_RE.search(sql):
        return sql
    return sql.rstrip().rstrip(";") + " RETURNING *;"


def detect_driver(conn) -> str:
    """Duck-type a connection and return its driver kind."""
    if conn is None:
        raise ValueError("detect_driver requires a connection")
    module = type(conn).__module__
    if module == "sqlite3":
        return "sqlite3"
    if module == "duckdb":
        return "duckdb"
    if hasattr(conn, "fetch") and hasattr(conn, "execute"):
        return "asyncpg"
    if module.startswith("psycopg"):
        return "psycopg"
    if module.startswith("pymysql"):
        return "pymysql"
    if module.startswith("mysql"):
        return "mysql"
    if hasattr(conn, "cursor"):
        return "dbapi"
    raise ValueError(
        f"unsupported database connection {type(conn).__name__!r}; "
        "pass driver=... to force a driver kind"
    )


class PipeqlDriver:
    """A database connection pre-wired with PipeQL compilation & dispatch."""

    def __init__(
        self,
        conn,
        *,
        dialect: Optional[str] = None,
        driver: Optional[str] = None,
    ):
        if conn is None:
            raise ValueError("create_pipeql_driver requires a database connection")
        self._conn = conn
        self.driver = driver or detect_driver(conn)
        self.dialect = dialect or _DIALECT_BY_DRIVER.get(self.driver)
        if self.dialect is None:
            raise ValueError(
                f"cannot infer a target dialect for driver {self.driver!r}; pass dialect=..."
            )
        self._cache: Dict[Tuple[str, str], Dict[str, Any]] = {}

    @staticmethod
    def _source_of(source: Any) -> str:
        """Resolve a builder object (duck-typed) or a raw source string."""
        if isinstance(source, str):
            return source
        if hasattr(source, "source") and callable(source.source):
            return source.source()
        if hasattr(source, "source"):
            return source.source
        raise TypeError(
            "source must be a PipeQL string or a builder object with .source()"
        )

    def _prepare(
        self,
        source: Any,
        params: Optional[Dict[str, Any]],
        whole_object_as_data: bool = False,
    ) -> Tuple[str, Dict[str, Any]]:
        """Resolve builders + ``$data`` object expansion into a source string."""
        builder_values = getattr(source, "values", None)
        source = self._source_of(source)
        merged = dict(params or {})
        if isinstance(builder_values, dict):
            merged.update(builder_values)
        if not _DATA_RE.search(source):
            return source, merged
        data = merged.get("data") if isinstance(merged, dict) else None
        if data is None and isinstance(merged, dict) and merged:
            data = merged
        expanded, values = _expand_data(source, data)
        merged.update(values)
        return expanded, merged

    def _compile(self, source: str) -> Dict[str, Any]:
        key = (self.dialect, source)
        hit = self._cache.get(key)
        if hit is None:
            hit = _compile(source, self.dialect)
            self._cache[key] = hit
        return hit

    def compile(
        self,
        source: str,
        params: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Compile ``source`` and bind params; adds the positional ``args`` list."""
        src, p = self._prepare(source, params)
        compiled = self._compile(src)
        return {**compiled, "args": _bind(compiled, p)}

    def _run_sync(self, compiled: Dict[str, Any], args: List[Any]) -> Dict[str, Any]:
        # select AND union return rows; mutations and DDL go through execute.
        # Keying off `statement_type == "select"` dropped rows from unions.
        is_select = compiled["statement_type"] in ("select", "union")
        cursor = self._conn.cursor()
        try:
            cursor.execute(compiled["sql"], args)
            if is_select:
                return {"rows": _rows_as_dicts(cursor)}
            return {
                "rows": [],
                "last_id": getattr(cursor, "lastrowid", None),
                "changes": getattr(cursor, "rowcount", None),
            }
        finally:
            try:
                cursor.close()
            except Exception:
                pass

    def _run_returning(self, sql: str, args: List[Any]) -> List[Dict[str, Any]]:
        """Run a row-returning statement (``... RETURNING *``) and fetch dict rows."""
        cursor = self._conn.cursor()
        try:
            cursor.execute(sql, args)
            return _rows_as_dicts(cursor)
        finally:
            try:
                cursor.close()
            except Exception:
                pass

    async def _run_async(self, compiled: Dict[str, Any], args: List[Any]) -> Dict[str, Any]:
        # select AND union are read-only — both must return rows.
        is_select = compiled["statement_type"] in ("select", "union")
        if is_select:
            rows = await self._conn.fetch(compiled["sql"], *args)
            return {"rows": rows}
        status = await self._conn.execute(compiled["sql"], *args)
        return {"rows": [], "last_id": None, "changes": _parse_command_tag(status)}

    def query(self, source: str, params: Optional[Dict[str, Any]] = None):
        """Run any statement; rows for SELECT, ``{last_id, changes, rows: []}`` for mutations."""
        if self.driver in _ASYNC_DRIVERS:
            raise TypeError(f"{self.driver} is async-only: use aquery()/aexecute()")
        src, p = self._prepare(source, params)
        compiled = self._compile(src)
        raw = self._run_sync(compiled, _bind(compiled, p))
        if compiled["statement_type"] in ("select", "union"):
            return raw["rows"]
        return {"last_id": raw["last_id"], "changes": raw["changes"], "rows": []}

    def execute(self, source: str, params: Optional[Dict[str, Any]] = None):
        """Run a statement; mutations return ``{last_id, changes, rows: []}``, selects ``{rows}``."""
        if self.driver in _ASYNC_DRIVERS:
            raise TypeError(f"{self.driver} is async-only: use aquery()/aexecute()")
        src, p = self._prepare(source, params)
        compiled = self._compile(src)
        raw = self._run_sync(compiled, _bind(compiled, p))
        if compiled["statement_type"] in ("select", "union"):
            return {"rows": raw["rows"]}
        return {"last_id": raw["last_id"], "changes": raw["changes"], "rows": []}

    def _fetch_impl(self, source: str, params: Optional[Dict[str, Any]], whole_object: bool):
        """Shared insert/update + return via RETURNING (fallback for MySQL)."""
        src, p = self._prepare(source, params, whole_object_as_data=whole_object)
        compiled = self._compile(src)
        args = _bind(compiled, p)
        if self.driver in _NO_RETURNING:
            raw = self._run_sync(compiled, args)
            return {"last_id": raw["last_id"], "changes": raw["changes"], "rows": []}
        rows = self._run_returning(_with_returning(compiled["sql"]), args)
        if not rows:
            return {"last_id": None, "changes": 0, "rows": []}
        return rows[0] if len(rows) == 1 else rows

    async def _afetch_impl(
        self, source: str, params: Optional[Dict[str, Any]], whole_object: bool
    ):
        src, p = self._prepare(source, params, whole_object_as_data=whole_object)
        compiled = self._compile(src)
        args = _bind(compiled, p)
        rows = await self._conn.fetch(_with_returning(compiled["sql"]), *args)
        if not rows:
            return {"last_id": None, "changes": 0, "rows": []}
        return rows[0] if len(rows) == 1 else rows

    def insert_and_fetch(self, source: str, params: Optional[Dict[str, Any]] = None):
        """Insert and return the created row(s).

        The entire ``params`` object is treated as the data object when the
        source uses ``$data``:
        ``db.insert_and_fetch("into notes | insert $data", req_body)``
        """
        if self.driver in _ASYNC_DRIVERS:
            raise TypeError(f"{self.driver} is async-only: use ainsert_and_fetch()")
        return self._fetch_impl(source, params, whole_object=True)

    def update_and_fetch(self, source: str, params: Optional[Dict[str, Any]] = None):
        """Update matching rows and return the updated row(s)."""
        if self.driver in _ASYNC_DRIVERS:
            raise TypeError(f"{self.driver} is async-only: use aupdate_and_fetch()")
        return self._fetch_impl(source, params, whole_object=False)

    async def ainsert_and_fetch(self, source: str, params: Optional[Dict[str, Any]] = None):
        """Async variant of :meth:`insert_and_fetch` for ``asyncpg``."""
        if self.driver not in _ASYNC_DRIVERS:
            raise TypeError(f"{self.driver} is synchronous: use insert_and_fetch()")
        return await self._afetch_impl(source, params, whole_object=True)

    async def aupdate_and_fetch(self, source: str, params: Optional[Dict[str, Any]] = None):
        """Async variant of :meth:`update_and_fetch` for ``asyncpg``."""
        if self.driver not in _ASYNC_DRIVERS:
            raise TypeError(f"{self.driver} is synchronous: use update_and_fetch()")
        return await self._afetch_impl(source, params, whole_object=False)

    async def aquery(self, source: str, params: Optional[Dict[str, Any]] = None):
        """Async variant of :meth:`query` for ``asyncpg`` connections."""
        if self.driver not in _ASYNC_DRIVERS:
            raise TypeError(f"{self.driver} is synchronous: use query()/execute()")
        src, p = self._prepare(source, params)
        compiled = self._compile(src)
        raw = await self._run_async(compiled, _bind(compiled, p))
        if compiled["statement_type"] in ("select", "union"):
            return raw["rows"]
        return {"last_id": raw["last_id"], "changes": raw["changes"], "rows": []}

    async def aexecute(self, source: str, params: Optional[Dict[str, Any]] = None):
        """Async variant of :meth:`execute` for ``asyncpg`` connections."""
        if self.driver not in _ASYNC_DRIVERS:
            raise TypeError(f"{self.driver} is synchronous: use query()/execute()")
        src, p = self._prepare(source, params)
        compiled = self._compile(src)
        raw = await self._run_async(compiled, _bind(compiled, p))
        if compiled["statement_type"] in ("select", "union"):
            return {"rows": raw["rows"]}
        return {"last_id": raw["last_id"], "changes": raw["changes"], "rows": []}

    def close(self):
        """Close the underlying connection (best-effort, respects async)."""
        closer = getattr(self._conn, "close", None)
        if callable(closer):
            return closer()


def create_pipeql_driver(conn, *, dialect: Optional[str] = None, driver: Optional[str] = None):
    """Wrap a database connection with PipeQL compilation and dispatch.

    Args:
        conn: A DB-API 2.0 connection (``sqlite3``, ``duckdb``, ``psycopg``,
            ``pymysql``, ...) or an ``asyncpg`` connection.
        dialect: Target compile dialect. Inferred from the driver when omitted.
        driver: Force a driver kind instead of duck-typed auto-detection.
    """
    return PipeqlDriver(conn, dialect=dialect, driver=driver)

File: go/driver.go
// Package pipeql provides zero-boilerplate database adapters for Go standard
// library's database/sql package.
//
// WrapDB wraps any *sql.DB connection (PostgreSQL, SQLite, DuckDB, MySQL) and
// auto-handles:
//   - compilation (PipeQL source -> dialect SQL)
//   - parameter binding (named PipeQL params map -> positional []any driver args)
//   - $data object expansion (partial insert/update with zero boilerplate)
//
// Usage:
//
//	db, _ := sql.Open("sqlite3", ":memory:")
//	pdb := pipeql.WrapDB(db, "sqlite")
//
//	rows, err := pdb.Query(ctx, "from users | filter age >= $min", map[string]any{"min": 18})
//	res, err := pdb.Exec(ctx, "into notes | insert [title = $title]", map[string]any{"title": "Note 1"})
//	res, err := pdb.Exec(ctx, "into notes | insert $data", map[string]any{"data": map[string]any{"title": "Note 2"}})
package pipeql

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"strings"
)

var (
	dataRegex  = regexp.MustCompile(`\$data\b`)
	identRegex = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)
)

// PipeqlDB wraps *sql.DB with PipeQL compilation and parameter binding.
type PipeqlDB struct {
	*sql.DB
	Dialect string
}

// PipeqlTx wraps *sql.Tx with PipeQL compilation and parameter binding.
type PipeqlTx struct {
	*sql.Tx
	Dialect string
}

// PipeqlConn wraps *sql.Conn with PipeQL compilation and parameter binding.
type PipeqlConn struct {
	*sql.Conn
	Dialect string
}

// WrapDB wraps a *sql.DB with the target PipeQL SQL dialect ("postgres", "sqlite", "duckdb", "mysql").
func WrapDB(db *sql.DB, dialect string) *PipeqlDB {
	if dialect == "" {
		dialect = "postgres"
	}
	return &PipeqlDB{DB: db, Dialect: dialect}
}

// WrapTx wraps a *sql.Tx with the target PipeQL SQL dialect.
func WrapTx(tx *sql.Tx, dialect string) *PipeqlTx {
	if dialect == "" {
		dialect = "postgres"
	}
	return &PipeqlTx{Tx: tx, Dialect: dialect}
}

// WrapConn wraps a *sql.Conn with the target PipeQL SQL dialect.
func WrapConn(conn *sql.Conn, dialect string) *PipeqlConn {
	if dialect == "" {
		dialect = "postgres"
	}
	return &PipeqlConn{Conn: conn, Dialect: dialect}
}

// QuerySource extracts the source string from a string or a *Query builder.
func querySourceOf(source any) (string, error) {
	switch s := source.(type) {
	case string:
		return s, nil
	case *Query:
		return s.Source(), nil
	case interface{ Source() string }:
		return s.Source(), nil
	default:
		return "", fmt.Errorf("pipeql: source must be a string or query builder, got %T", source)
	}
}

func expandData(source string, data map[string]any) (string, map[string]any, error) {
	if len(data) == 0 {
		return "", nil, fmt.Errorf("pipeql: $data object expansion requires at least one property")
	}

	// Sort keys for deterministic column ordering
	keys := make([]string, 0, len(data))
	for k := range data {
		if !identRegex.MatchString(k) {
			return "", nil, fmt.Errorf("pipeql: cannot expand $data: column %q is not a valid identifier", k)
		}
		keys = append(keys, k)
	}
	sort.Strings(keys)

	parts := dataRegex.Split(source, -1)
	var chunks []string
	chunks = append(chunks, parts[0])
	values := make(map[string]any)
	n := 0

	for i := 1; i < len(parts); i++ {
		prev := strings.TrimRight(chunks[len(chunks)-1], " \t\r\n")
		var last byte
		if len(prev) > 0 {
			last = prev[len(prev)-1]
		}
		inBrackets := last == '[' || last == ','

		var assignments []string
		for _, key := range keys {
			pname := fmt.Sprintf("data%d", n)
			n++
			values[pname] = data[key]
			assignments = append(assignments, fmt.Sprintf("%s = $%s", key, pname))
		}
		body := strings.Join(assignments, ", ")
		if inBrackets {
			chunks = append(chunks, body)
		} else {
			chunks = append(chunks, "["+body+"]")
		}
		chunks = append(chunks, parts[i])
	}

	return strings.Join(chunks, ""), values, nil
}

// CompileAndBind prepares source and binds named params into positional driver arguments.
func CompileAndBind(source any, dialect string, params map[string]any) (string, []any, *Result, error) {
	src, err := querySourceOf(source)
	if err != nil {
		return "", nil, nil, err
	}

	mergedParams := make(map[string]any, len(params))
	for k, v := range params {
		mergedParams[k] = v
	}

	if dataRegex.MatchString(src) {
		dataObj, hasData := mergedParams["data"]
		var dataMap map[string]any
		if hasData {
			if m, ok := dataObj.(map[string]any); ok {
				dataMap = m
			}
		}
		if dataMap == nil {
			dataMap = mergedParams
		}
		expandedSrc, dataVals, err := expandData(src, dataMap)
		if err != nil {
			return "", nil, nil, err
		}
		src = expandedSrc
		for k, v := range dataVals {
			mergedParams[k] = v
		}
	}

	res, err := Compile(src, dialect)
	if err != nil {
		return "", nil, nil, err
	}

	// Extract parameter types from analysis
	type paramMeta struct {
		Name string `json:"name"`
		Ty   string `json:"ty"`
	}
	type analysisDoc struct {
		ParamMap []paramMeta `json:"param_map"`
	}
	var analysis analysisDoc
	if len(res.Analysis) > 0 {
		_ = json.Unmarshal(res.Analysis, &analysis)
	}
	paramTypes := make(map[string]string, len(analysis.ParamMap))
	for _, p := range analysis.ParamMap {
		paramTypes[p.Name] = p.Ty
	}

	args := make([]any, 0, len(res.Params))
	for _, name := range res.Params {
		if val, exists := mergedParams[name]; exists {
			args = append(args, val)
		} else if paramTypes[name] == "any" || paramTypes[name] == "Any" {
			return "", nil, nil, fmt.Errorf("pipeql: missing value for parameter $%s — pass in params map", name)
		} else {
			// Literal-derived parameter (in select/filter)
			args = append(args, name)
		}
	}

	return res.SQL, args, res, nil
}

// Query executes a PipeQL query against the database, returning rows.
func (db *PipeqlDB) Query(ctx context.Context, source any, params map[string]any) (*sql.Rows, error) {
	sqlStr, args, _, err := CompileAndBind(source, db.Dialect, params)
	if err != nil {
		return nil, err
	}
	return db.DB.QueryContext(ctx, sqlStr, args...)
}

// QueryRow executes a PipeQL query and returns a single row.
func (db *PipeqlDB) QueryRow(ctx context.Context, source any, params map[string]any) (*sql.Row, error) {
	sqlStr, args, _, err := CompileAndBind(source, db.Dialect, params)
	if err != nil {
		return nil, err
	}
	return db.DB.QueryRowContext(ctx, sqlStr, args...), nil
}

// Exec executes a PipeQL mutation/DDL statement and returns sql.Result.
func (db *PipeqlDB) Exec(ctx context.Context, source any, params map[string]any) (sql.Result, error) {
	sqlStr, args, _, err := CompileAndBind(source, db.Dialect, params)
	if err != nil {
		return nil, err
	}
	return db.DB.ExecContext(ctx, sqlStr, args...)
}

// BeginTx starts a transaction wrapped with PipeQL capabilities.
func (db *PipeqlDB) BeginTx(ctx context.Context, opts *sql.TxOptions) (*PipeqlTx, error) {
	tx, err := db.DB.BeginTx(ctx, opts)
	if err != nil {
		return nil, err
	}
	return &PipeqlTx{Tx: tx, Dialect: db.Dialect}, nil
}

// Query executes a PipeQL query in a transaction.
func (tx *PipeqlTx) Query(ctx context.Context, source any, params map[string]any) (*sql.Rows, error) {
	sqlStr, args, _, err := CompileAndBind(source, tx.Dialect, params)
	if err != nil {
		return nil, err
	}
	return tx.Tx.QueryContext(ctx, sqlStr, args...)
}

// QueryRow executes a PipeQL query in a transaction and returns a single row.
func (tx *PipeqlTx) QueryRow(ctx context.Context, source any, params map[string]any) (*sql.Row, error) {
	sqlStr, args, _, err := CompileAndBind(source, tx.Dialect, params)
	if err != nil {
		return nil, err
	}
	return tx.Tx.QueryRowContext(ctx, sqlStr, args...), nil
}

// Exec executes a PipeQL mutation/DDL statement in a transaction.
func (tx *PipeqlTx) Exec(ctx context.Context, source any, params map[string]any) (sql.Result, error) {
	sqlStr, args, _, err := CompileAndBind(source, tx.Dialect, params)
	if err != nil {
		return nil, err
	}
	return tx.Tx.ExecContext(ctx, sqlStr, args...)
}

// Query executes a PipeQL query on a connection.
func (c *PipeqlConn) Query(ctx context.Context, source any, params map[string]any) (*sql.Rows, error) {
	sqlStr, args, _, err := CompileAndBind(source, c.Dialect, params)
	if err != nil {
		return nil, err
	}
	return c.Conn.QueryContext(ctx, sqlStr, args...)
}

// Exec executes a PipeQL mutation/DDL statement on a connection.
func (c *PipeqlConn) Exec(ctx context.Context, source any, params map[string]any) (sql.Result, error) {
	sqlStr, args, _, err := CompileAndBind(source, c.Dialect, params)
	if err != nil {
		return nil, err
	}
	return c.Conn.ExecContext(ctx, sqlStr, args...)
}
